esi_leap/objects/offer.py

```diff
@@ -78,9 +82,31 @@ def get_availabilities(self):
 
     def create(self, context=None):
         updates = self.obj_get_changes()
+
+        updates['uuid'] = uuidutils.generate_uuid()
+
+        if 'start_time' not in updates:
+            updates['start_time'] = datetime.datetime.now()
+        if 'end_time' not in updates:
+            updates['end_time'] = datetime.datetime.max
+
+        if updates['start_time'] >= updates['end_time']:
+            raise exception.\
+                InvalidTimeRange(resource="an offer",
+                                 start_time=str(updates['start_time']),
+                                 end_time=str(updates['end_time']))
+
         db_offer = self.dbapi.offer_create(updates)
         self._from_db_object(context, self, db_offer)
 
+    def cancel(self):
+        self.status = statuses.CANCELLED
+        self.save(None)
+
+        contracts = self.dbapi.contract_get_all({'owner': self.uuid})
+        for c in contracts:
+            c.cancel()
```

Comment:
I think we may want to cancel contracts before setting the status to cancelled; this way, if there's an error in canceling a contract for whatever reason, the offer is marked as still open.

Reply:
I would also have to find a way to rollback the cancelling of the contracts that did get cancelled for whatever reason, wouldnt I?

Reply:
Seems like a whole can of worms potentially

Reply:
Yep, and it's probably something we want to handle in a future update; in the meantime, I'd argue that if the contracts can't all be canceled, then the offer's status should not be updated.

Reply:
I have made this change.